lib/ecto/multi.ex

```diff
@@ -0,0 +1,364 @@
+  that can be used for introspection.
+
+  ## Changesets
+
+  If Multi contains operations that accept changesets (like `insert/4`,
+  `update/4` or `delete/4`) they will be checked before starting transaction
+  if any of them isn't invalid. In case any has errors, the transaction won't
+  be even started and the error will be immediately returned.
+
+  ## Run
+
+  Multi allows you to run arbitrary functions as part of your transaction via
+  the `run/3` and `run/5`. Those functions will receive changes so far as the
+  first argument and have to return `{:ok, value}` or `{:error, value}` as
+  their result. Returning an error will abort any further operations and
+  make the whole multi fail.
+
+  ## Example
+
+  Let's look at an example definition and usage. The use case we'll be
+  looking into is resetting password. We need to update the account
+  with proper information, log the request and remove all current sessions.
+  We define a function creating the Multi structure probably in some sort of
+  service layer:
+
+      defmodule Service do
+        alias Ecto.Multi
+        import Ecto
+
+        def password_reset(account, params) do
+          Multi.new
+          |> Multi.update(:account, Account.password_reset_changeset(account, params))
+          |> Multi.insert(:log, Log.password_reset_changeset(account, params))
+          |> Multi.delete_all(:sessions, assoc(account, :sessions))
+        end
+      end
+
+  We can later execute it in the integration layer using Repo:
+
+      Repo.transaction(Service.password_reset(account, params))
+
+  By pattern matching on the result we can differentiate different conditions:
+
+      case result do
+        {:ok, %{account: account, log: log, sessions: sessions}} ->
+          # Operation was successful, we can access results (exactly the same
+          # we would get from running corresponding Repo functions)
+          # under keys we used for naming the operations.
+        {:error, failed_operation, failed_value, changes_so_far} ->
+          # One of the operations failed. We can access the operation's failure
+          # value (like changeset for operations on changesets) to prepare
+          # proper response. We also get access to results of any operations
+          # that succeeded before the indicated operation failed.
+      end
+
+  We can also easily unit test our transaction without actually running it.
+  Since changesets can use in-memory-data, we can use an account that is
+  constructed in memory as well (without persisting it to the database):
+
+      test "dry run password_reset" do
+        account = %Account{password: "letmein"}
+        multi = Service.password_reset(account, params)
+
+        assert [
+          {:account, {:update, account_changeset, []}},
+          {:log, {:insert, log_changeset, []}},
+          {:sessions, {:delete_all, query, []}}
+        ] = Ecto.Multi.to_list(multi)
```

Comment:
Oh my bad. I didn’t notice that it was just one `=` :) Yeah I think an example would be helpful. I was wondering where the account_changeset, log_changeset, etc. came from because I didn’t notice the pattern match.

> On Feb 18, 2016, at 11:47 AM, Michał Muskała <notifications@github.com> wrote:
> 
> In lib/ecto/multi.ex <https://github.com/elixir-lang/ecto/pull/1138#discussion_r53342854>:
> 
> > +          # that succeeded before the indicated operation failed.
> > +      end
> > +
> > +  We can also easily unit test our transaction without actually running it.
> > +  Since changesets can use in-memory-data, we can use an account that is
> > +  constructed in memory as well (without persisting it to the database):
> > +
> > +      test "dry run password_reset" do
> > +        account = %Account{password: "letmein"}
> > +        multi = Service.password_reset(account, params)
> > +
> > +        assert [
> > +          {:account, {:update, account_changeset, []}},
> > +          {:log, {:insert, log_changeset, []}},
> > +          {:sessions, {:delete_all, query, []}}
> > +        ] = Ecto.Multi.to_list(multi)
> It's using = to pattern match and assign parts of the value to variables to assert later. It's not possible with ==.
> 
> Maybe an example of that later will help with why that's done.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/elixir-lang/ecto/pull/1138/files#r53342854>.
>